addons/account/tests/test_account_bank_statement.py

```diff
@@ -824,597 +822,107 @@ def test_statement_line_move_onchange_1(self):
-        :param amount_currency:             The amount in the foreign currency.
-        :param counterpart_amount:          The amount of the invoice to reconcile.
-        :param journal_currency:            The journal's currency as a res.currency record.
-        :param foreign_currency:            The foreign currency as a res.currency record.
-        :param counterpart_currency:        The invoice currency as a res.currency record.
-        :param expected_liquidity_values:   The expected account.move.line values for the liquidity line.
-        :param expected_counterpart_values: The expected account.move.line values for the counterpart line.
-        '''
-        if journal_currency:
-            journal.currency_id = journal_currency.id
-
-        statement = self.env['account.bank.statement'].create({
-            'name': 'test_statement',
-            'date': '2019-01-01',
-            'journal_id': journal.id,
-            'line_ids': [
-                (0, 0, {
-                    'date': '2019-01-01',
-                    'payment_ref': 'line_1',
-                    'partner_id': self.partner_a.id,
-                    'foreign_currency_id': foreign_currency and foreign_currency.id,
-                    'amount': amount,
-                    'amount_currency': amount_currency,
-                }),
-            ],
-        })
-        statement_line = statement.line_ids
-
-        # - There is 3 flows to check:
-        #   * The invoice will fully reconcile the statement line.
-        #   * The invoice will partially reconcile the statement line and leads to an open balance in debit.
-        #   * The invoice will partially reconcile the statement line and leads to an open balance in credit.
-        # - The dates are different to be sure the reconciliation will preserve the conversion rate bank side.
-        move_type = 'out_invoice' if counterpart_amount < 0.0 else 'in_invoice'
-
-        test_invoices = self.env['account.move'].create([
-            {
-                'move_type': move_type,
-                'invoice_date': fields.Date.from_string('2016-01-01'),
-                'date': fields.Date.from_string('2016-01-01'),
-                'partner_id': self.partner_a.id,
-                'currency_id': counterpart_currency.id,
-                'invoice_line_ids': [
-                    (0, None, {
-                        'name': 'counterpart line, same amount',
-                        'account_id': self.company_data['default_account_revenue'].id,
-                        'quantity': 1,
-                        'price_unit': abs(counterpart_amount),
+    def test_prepare_counterpart_amounts_using_st_line_rate(self):
+
+        def assertAppliedRate(
+                journal_currency, foreign_currency, aml_currency,
+                amount, amount_currency, aml_amount_currency, aml_balance,
+                expected_amount_currency, expected_balance,
+        ):
```

Comment:
Might be nice to do this in a [subtest](https://docs.python.org/3/library/unittest.html#unittest.TestCase.subTest)?